### PR TITLE
Porting Weave Layer to Torch

### PR DESCRIPTION
**Pull request by NimishaDey**
## Description

This PR is to port Weave Layer to PyTorch.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [X] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [X] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [X] Run `mypy -p deepchem` and check no errors
  - [X] Run `flake8 <modified file> --count` and check no errors
  - [X] Run `python -m doctest <modified file>` and check no errors
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -2972,3 +2972,339 @@ def _dynamic_partition(self, input_tensor: torch.Tensor,
         partitions = [input_tensor[mask] for mask in partition_masks]
 
         return partitions
+
+
+class WeaveLayer(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:2977]: Can you add the unit tests to this PR?

Comment (NimishaDey) [reply deepchem/models/torch_models/layers.py:2977]: I have added the unit tests.

Comment (rbharath) [inline deepchem/models/torch_models/layers.py:2977]: Can you add this class to the layers.rst file in the docs? We should make sure it appears in the docs for layers

**Review by rbharath — COMMENTED** [review]
I've made a request to add to the docs and turned on the CI. Let's see if everything passes correctly

File: deepchem/models/torch_models/layers.py (hunk 1 above)
Comment (NimishaDey) [reply deepchem/models/torch_models/layers.py:2977]: I have now added WeaveLayer to docs.

**Review by rbharath — COMMENTED** [review]
@NimishaDey Can you rebase this PR? I think it's ready to merge, but we will need a rebase and then to rerun the test suite

File: docs/source/api_reference/layers.rst [hunk 2]
```diff
@@ -207,6 +207,9 @@ Torch Layers
 .. autoclass:: deepchem.models.torch_models.layers.DTNNEmbedding
   :members:
 
+.. autoclass:: deepchem.models.torch_models.layers.WeaveLayer
```
Comment (rbharath) [inline docs/source/api_reference/layers.rst:210]: Please make sure to add to https://github.com/deepchem/deepchem/blob/master/docs/source/api_reference/torch_layers.csv as well

Comment (NimishaDey) [reply docs/source/api_reference/layers.rst:210]: Sir I have added it now.

**Pull request merged**